File: app/src/main/java/sample/mail/ru/httpsample/MainActivity.java
package sample.mail.ru.httpsample;

import android.app.Fragment;
import android.content.SharedPreferences;
import android.os.AsyncTask;
import android.os.Bundle;
import android.preference.PreferenceFragment;
import android.preference.PreferenceManager;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.Menu;
import android.view.MenuItem;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import java.io.BufferedInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.lang.ref.WeakReference;
import java.net.Socket;
import java.net.UnknownHostException;
import java.nio.charset.Charset;
import java.util.Arrays;

import javax.net.SocketFactory;
import javax.net.ssl.HostnameVerifier;
import javax.net.ssl.HttpsURLConnection;
import javax.net.ssl.SSLHandshakeException;
import javax.net.ssl.SSLSession;
import javax.net.ssl.SSLSocket;
import javax.net.ssl.SSLSocketFactory;

public class MainActivity extends AppCompatActivity {

    interface RequestListener {
        void onRequestResult(String result);
        void onRequestError(int errorStringID);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        getFragmentManager().beginTransaction()
                .replace(R.id.content_frame, new MainFragment())
                .commit();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();

        if (id == R.id.action_settings) {
            openSettings();
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    public void openSettings() {
        getFragmentManager().beginTransaction()
                .replace(R.id.content_frame, new SettingsFragment())
                .addToBackStack("SETTINGS")
                .commit();
    }

    @Override
    public void onBackPressed() {
        if (getFragmentManager().getBackStackEntryCount() > 0) {
            getFragmentManager().popBackStackImmediate();
        }
        else {
            super.onBackPressed();
        }
    }

    public static class MainFragment extends Fragment implements RequestListener {
        private Button mOverHttp;
        private Button mOverSocket;
        private Button mDownloadJSON;

        private TextView mOutput;

        private AsyncTask<?, ?, ?> mRequestTask;

        @Nullable
        @Override
        public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
            View v = inflater.inflate(R.layout.content_main, null);
            mOverHttp = (Button)v.findViewById(R.id.http_button);
            mOverHttp.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (mRequestTask != null) {
                        mRequestTask.cancel(true);
                    }
                    mOutput.setText("");
                    mRequestTask = new HttpAsyncRequest(MainFragment.this).execute(
                            isOverHTTPS() ? "https://mail.ru" : "http://mail.ru"
                    );
                }
            });
            mOverSocket = (Button)v.findViewById(R.id.socket_button);
            mOverSocket.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (mRequestTask != null) {
                        mRequestTask.cancel(true);
                    }
                    mOutput.setText("");
                    mRequestTask = new SocketAsyncRequest(MainFragment.this).execute(
                            "mail.ru", isOverHTTPS() ? "443" : "80"
                    );
                }
            });
            mDownloadJSON = (Button)v.findViewById(R.id.download_json);
            mDownloadJSON.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (mRequestTask != null) {
                        mRequestTask.cancel(true);
                    }
                    mOutput.setText("");
                    mRequestTask = new DownloadJSONAsync(MainFragment.this).execute(DownloadJSONAsync.JSON_URL);
                }
            });

            mOutput = (TextView)v.findViewById(R.id.output);

            return v;
        }

        @Override
        public void onStart() {
            super.onStart();
            mOverHttp.setText(getString(isOverHTTPS() ? R.string.https : R.string.http));
        }

        @Override
        public void onStop() {
            super.onStop();
            if (mRequestTask != null) {
                mRequestTask.cancel(true);
            }
        }

        @Override
        public void onRequestResult(String result) {
            mOutput.setText(result);
        }

        @Override
        public void onRequestError(int errorStringID) {
            mOutput.setText(errorStringID);
        }

        private boolean isOverHTTPS() {
            SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(getActivity());
            return prefs.getBoolean("pref_https", true);
        }
    }

    public static class SettingsFragment extends PreferenceFragment {
        @Override
        public void onCreate(Bundle savedInstanceState) {
            super.onCreate(savedInstanceState);
            addPreferencesFromResource(R.xml.preferences);
        }
    }
}

File: app/src/main/java/sample/mail/ru/httpsample/MainActivity.kt
package sample.mail.ru.httpsample

import android.os.Bundle
import android.preference.PreferenceManager
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.view.*
import android.widget.Button
import android.widget.TextView
import kotlinx.coroutines.*
import sample.mail.ru.httpsample.request.DownloadJSONAsync
import sample.mail.ru.httpsample.request.HttpAsyncRequest
import sample.mail.ru.httpsample.request.RequestExecutor
import sample.mail.ru.httpsample.request.SocketAsyncRequest

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)

        if (savedInstanceState == null) {
            supportFragmentManager.beginTransaction()
                    .replace(R.id.content_frame, MainFragment())
                    .commit()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId

        if (id == R.id.action_settings) {
            openSettings()
            return true
        }

        return super.onOptionsItemSelected(item)
    }

    private fun openSettings() {
        supportFragmentManager.beginTransaction()
                .replace(R.id.content_frame, SettingsFragment())
                .addToBackStack("SETTINGS")
                .commit()
    }

    override fun onBackPressed() {
        if (supportFragmentManager.backStackEntryCount > 0) {
            supportFragmentManager.popBackStackImmediate()
        } else {
            super.onBackPressed()
        }
    }

    class MainFragment : android.support.v4.app.Fragment(), View.OnClickListener {
        private lateinit var overHttp: Button
        private lateinit var overSocket: Button
        private lateinit var downloadJSON: Button

        private lateinit var output: TextView

        private var job : Job? = null

        private val overSSL: Boolean
            get() {
                val prefs = PreferenceManager.getDefaultSharedPreferences(activity)
                return prefs.getBoolean("pref_https", true)
            }

        override fun onStart() {
            super.onStart()
            overHttp.text = getString(if (overSSL) R.string.https else R.string.http)
        }

        override fun onStop() {
            job?.cancel()
            super.onStop()
        }

        override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
            val v = inflater.inflate(R.layout.content_main, container, false)

            overHttp = v.findViewById<View>(R.id.http_button) as Button
            overSocket = v.findViewById<View>(R.id.socket_button) as Button
            downloadJSON = v.findViewById<View>(R.id.download_json) as Button
            output = v.findViewById<View>(R.id.output) as TextView

            overHttp.setOnClickListener(this)
            overSocket.setOnClickListener(this)
            downloadJSON.setOnClickListener(this)

            return v
        }

        override fun onClick(v: View?) {
            when(v?.id) {
                R.id.http_button -> requestToView(HttpAsyncRequest(if (overSSL) Consts.HTTPS + Consts.DEFAULT_URL else Consts.HTTP + Consts.DEFAULT_URL))
                R.id.socket_button -> requestToView(SocketAsyncRequest(Consts.DEFAULT_URL, if (overSSL) 443 else 80))
                R.id.download_json -> requestToView(DownloadJSONAsync(Consts.JSON_URL))
            }
        }

        private fun requestToView(task : RequestExecutor) {
            output.text = ""
            job = GlobalScope.launch(Dispatchers.Main) {
                val result = async {
                    task.execute()
                }
                showRequestResult(result.await())
            }
        }

        private fun showRequestResult(request : Pair<String, Int>) {
            output.text = if (request.second > 0) getString(request.second) else request.first
        }
    }

    class SettingsFragment : android.support.v7.preference.PreferenceFragmentCompat() {
        override fun onCreatePreferences(savedInstanceState: Bundle?, param: String?) {
            addPreferencesFromResource(R.xml.preferences)
        }

    }
}
